pam-1.1.1/modules/pam_mkhomedir/pam_mkhomedir.8
.\"     Title: pam_mkhomedir
.\"    Author: [see the "AUTHOR" section]
.\" Generator: DocBook XSL Stylesheets v1.74.0 <http://docbook.sf.net/>
.\"      Date: 06/16/2009
.\"    Manual: Linux-PAM Manual
.\"    Source: Linux-PAM Manual
.\"  Language: English
.\"
.TH "PAM_MKHOMEDIR" "8" "06/16/2009" "Linux-PAM Manual" "Linux-PAM Manual"
.\" -----------------------------------------------------------------
.\" * (re)Define some macros
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" toupper - uppercase a string (locale-aware)
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.de toupper
.tr aAbBcCdDeEfFgGhHiIjJkKlLmMnNoOpPqQrRsStTuUvVwWxXyYzZ
\\$*
.tr aabbccddeeffgghhiijjkkllmmnnooppqqrrssttuuvvwwxxyyzz
..
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" SH-xref - format a cross-reference to an SH section
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.de SH-xref
.ie n \{\
.\}
.toupper \\$*
.el \{\
\\$*
.\}
..
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" SH - level-one heading that works better for non-TTY output
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.de1 SH
.\" put an extra blank line of space above the head in non-TTY output
.if t \{\
.sp 1
.\}
.sp \\n[PD]u
.nr an-level 1
.set-an-margin
.nr an-prevailing-indent \\n[IN]
.fi
.in \\n[an-margin]u
.ti 0
.HTML-TAG ".NH \\n[an-level]"
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
\." make the size of the head bigger
.ps +3
.ft B
.ne (2v + 1u)
.ie n \{\
.\" if n (TTY output), use uppercase
.toupper \\$*
.\}
.el \{\
.nr an-break-flag 0
.\" if not n (not TTY), use normal case (not uppercase)
\\$1
.in \\n[an-margin]u
.ti 0
.\" if not n (not TTY), put a border/line under subheading
.sp -.6
\l'\n(.lu'
.\}
..
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" SS - level-two heading that works better for non-TTY output
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.de1 SS
.sp \\n[PD]u
.nr an-level 1
.set-an-margin
.nr an-prevailing-indent \\n[IN]
.fi
.in \\n[IN]u
.ti \\n[SN]u
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.ps \\n[PS-SS]u
\." make the size of the head bigger
.ps +2
.ft B
.ne (2v + 1u)
.if \\n[.$] \&\\$*
..
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" BB/BE - put background/screen (filled box) around block of text
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.de BB
.if t \{\
.sp -.5
.br
.in +2n
.ll -2n
.gcolor red
.di BX
.\}
..
.de EB
.if t \{\
.if "\\$2"adjust-for-leading-newline" \{\
.sp -1
.\}
.br
.di
.in
.ll
.gcolor
.nr BW \\n(.lu-\\n(.i
.nr BH \\n(dn+.5v
.ne \\n(BHu+.5v
.ie "\\$2"adjust-for-leading-newline" \{\
\M[\\$1]\h'1n'\v'+.5v'\D'P \\n(BWu 0 0 \\n(BHu -\\n(BWu 0 0 -\\n(BHu'\M[]
.\}
.el \{\
\M[\\$1]\h'1n'\v'-.5v'\D'P \\n(BWu 0 0 \\n(BHu -\\n(BWu 0 0 -\\n(BHu'\M[]
.\}
.in 0
.sp -.5v
.nf
.BX
.in
.sp .5v
.fi
.\}
..
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" BM/EM - put colored marker in margin next to block of text
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.de BM
.if t \{\
.br
.ll -2n
.gcolor red
.di BX
.\}
..
.de EM
.if t \{\
.br
.di
.ll
.gcolor
.nr BH \\n(dn
.ne \\n(BHu
\M[\\$1]\D'P -.75n 0 0 \\n(BHu -(\\n[.i]u - \\n(INu - .75n) 0 0 -\\n(BHu'\M[]
.in 0
.nf
.BX
.in
.fi
.\}
..
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "Name"
pam_mkhomedir \- PAM module to create users home directory
.SH "Synopsis"
.fam C
.HP \w'\fBpam_mkhomedir\&.so\fR\ 'u
\fBpam_mkhomedir\&.so\fR [silent] [umask=\fImode\fR] [skel=\fIskeldir\fR]
.fam
.SH "DESCRIPTION"
.PP
The pam_mkhomedir PAM module will create a users home directory if it does not exist when the session begins\&. This allows users to be present in central database (such as NIS, kerberos or LDAP) without using a distributed file system or pre\-creating a large number of directories\&. The skeleton directory (usually
\FC/etc/skel/\F[]) is used to copy default files and also sets a umask for the creation\&.
.PP
The new users home directory will not be removed after logout of the user\&.
.SH "OPTIONS"
.PP
\fBsilent\fR
.RS 4
Don\'t print informative messages\&.
.RE
.PP
\fBumask=\fR\fB\fImask\fR\fR
.RS 4
The user file\-creation mask is set to
\fImask\fR\&. The default value of mask is 0022\&.
.RE
.PP
\fBskel=\fR\fB\fI/path/to/skel/directory\fR\fR
.RS 4
Indicate an alternative
\FCskel\F[]
directory to override the default
\FC/etc/skel\F[]\&.
.RE
.SH "MODULE TYPES PROVIDED"
.PP
Only the
\fBsession\fR
module type is provided\&.
.SH "RETURN VALUES"
.PP
PAM_BUF_ERR
.RS 4
Memory buffer error\&.
.RE
.PP
PAM_CRED_INSUFFICIENT
.RS 4
Insufficient credentials to access authentication data\&.
.RE
.PP
PAM_PERM_DENIED
.RS 4
Not enough permissions to create the new directory or read the skel directory\&.
.RE
.PP
PAM_USER_UNKNOWN
.RS 4
User not known to the underlying authentication module\&.
.RE
.PP
PAM_SUCCESS
.RS 4
Environment variables were set\&.
.RE
.SH "FILES"
.PP
\FC/etc/skel\F[]
.RS 4
Default skel directory
.RE
.SH "EXAMPLES"
.PP
A sample /etc/pam\&.d/login file:
.sp
.if n \{\
.RS 4
.\}
.fam C
.ps -1
.nf
.if t \{\
.sp -1
.\}
.BB lightgray adjust-for-leading-newline
.sp -1

  auth       requisite   pam_securetty\&.so
  auth       sufficient  pam_ldap\&.so
  auth       required    pam_unix\&.so
  auth       required    pam_nologin\&.so
  account    sufficient  pam_ldap\&.so
  account    required    pam_unix\&.so
  password   required    pam_unix\&.so
  session    required    pam_mkhomedir\&.so skel=/etc/skel/ umask=0022
  session    required    pam_unix\&.so
  session    optional    pam_lastlog\&.so
  session    optional    pam_mail\&.so standard
      
.EB lightgray adjust-for-leading-newline
.if t \{\
.sp 1
.\}
.fi
.fam
.ps +1
.if n \{\
.RE
.\}
.sp
.SH "SEE ALSO"
.PP

\fBpam.d\fR(5),
\fBpam\fR(8)\&.
.SH "AUTHOR"
.PP
pam_mkhomedir was written by Jason Gunthorpe <jgg@debian\&.org>\&.


noflushd/NEWS
* noflushd 2.7.5 (09. May 2005)

  - Fix handling of raid and similar meta devices on 2.6 kernels.

* noflushd 2.7.4 (08. Aug 2004)

  - Fix lousy spindown times on 2.6 kernels.
  - Fix SCSI support on 2.6 kernels.
  - Fix pid detection in generic startup script.
  - Document quirks with hotplugged drives.

* noflushd 2.7.3 (09. Apr 2004)

  - Fix race condition that might leave kupdated/pdflushd disabled
    aftern noflushd has terminated.
  - Fix spindown when all disks are auto-detected.

* noflushd 2.7.2 (16. Mar 2004)

  - A change in 2.7.1 erroneously flushed the buffer cache before
    spindown, causing frequent spinups. This change has been reverted.
    
* noflushd 2.7.1 (07. Mar 2004)

  - Fix immediate spinup on 2.6 kernels.
  - Elevate SCSI support from compile-time to command line option.
  - Bug fixes in the disk statistics parser for 2.6 kernels.
  - Ship kernel patch to track down which processes cause spinups.

* noflushd 2.7 (20. Jan 2004)

  - Support Linux kernel 2.6.x.
  - Default to monitorring all disks unless specific disks are given
    on the command line. This change should yield a sane default
    behaviour for a majority of users without requiring explicit
    configuration.

* noflushd 2.6.3 (13. May 2002)

  - Bugfix for arm and other archs with weird /proc/partitions layout.

* noflushd 2.6.2 (20. Apr 2002)

  - Add IDE raid support.
  - Add timeout logging script.

* noflushd 2.6.1 (03. Feb. 2002)

  - Fix BUG when using noflushd with disks on second IDE channel on 2.4
    kernels.

* noflushd 2.6 (22. Jan 2002)

  - Document noflushd's suckage with any journaling fs, including ext3.
  - New feature prevents spindown if machine is in interactive use.
  - Add hack fix some '0 spindown time' uglies.

* noflushd 2.5 (06. Aug 2001)

  - Make sure hotplugged disks get synced properly.
  - Only sync present partitions. (Fixes high CPU usage bug on devfs
    systems.)
  - Generic startup skript works on latest 2.4 kernels.

* noflushd 2.4 (13. Mar 2001)

  - Timeout lists are back to a working state.
  - Support very latest 2.4 kernels.

* noflushd 2.3 (12. Jan 2001)

  - Support latest 2.4 kernels. (Leaves kernels 2.3.99pre8 to 2.4.0-test8
    unsupported.)
  - Fix multi-disk support with long bdflush timeouts.

* noflushd 2.2 (22. Dec 2000)

  - Fixed bug in spindown code. Now actually spins down. Duh!

* noflushd 2.1 (20. Dec 2000)

  - Disks get synced prior to spindown.
  - Improved devfs support.

* noflushd 2.0 (14. Nov 2000)

  - noflushd now comes with its own syncer that takes over the task from
    kupdate when the first read-write mounted disk is spun down. This
    enhances spindown times in a multidisk environment at the expense of
    less smooth and more bursty background flushing when the noflushd 
    syncer is in use.

  - Conflicting entries in 2.2-style /proc/stat (e.g. /dev/sda and /dev/hda)
    won't cause noflushd to give up anymore. Instead, noflushd will treat
    both disks pretty much like one, i.e. they're either both up or spun
    down, no matter if activity only goes to one of them.

  - Compilation environment switched to automake/autoconf.

  - When installing via 'make install', noflushd is no longer added to
    any runlevel. 

  - noflushd moved to SourceForge. The project homepage is located at
    http://noflushd.sourceforge.net, a mailing list is available via
    noflushd-devel@lists.sourceforge.net. Visit the homepage for 
    subscription information.



pam-0.76/Linux-PAM/modules/pam_mkhomedir/pam_mkhomedir.c
/* PAM Make Home Dir module

   This module will create a users home directory if it does not exist
   when the session begins. This allows users to be present in central
   database (such as nis, kerb or ldap) without using a distributed
   file system or pre-creating a large number of directories.
   
   Here is a sample /etc/pam.d/login file for Debian GNU/Linux
   2.1:
   
   auth       requisite  pam_securetty.so
   auth       sufficient pam_ldap.so
   auth       required   pam_pwdb.so
   auth       optional   pam_group.so
   auth       optional   pam_mail.so
   account    requisite  pam_time.so
   account    sufficient pam_ldap.so
   account    required   pam_pwdb.so
   session    required   pam_mkhomedir.so skel=/etc/skel/ umask=0022
   session    required   pam_pwdb.so
   session    optional   pam_lastlog.so
   password   required   pam_pwdb.so   
   
   Released under the GNU LGPL version 2 or later
   Originally written by Jason Gunthorpe <jgg@debian.org> Feb 1999
   Structure taken from pam_lastlogin by Andrew Morgan 
     <morgan@parc.power.net> 1996
 */

/* I want snprintf dammit */
#define _GNU_SOURCE 1
#include <stdarg.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>
#include <pwd.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <dirent.h>

/*
 * here, we make a definition for the externally accessible function
 * in this file (this definition is required for static a module
 * but strongly encouraged generally) it is used to instruct the
 * modules include file to define the function prototypes.
 */

#define PAM_SM_SESSION

#include <security/pam_modules.h>
#include <security/_pam_macros.h>

/* argument parsing */
#define MKHOMEDIR_DEBUG      020	/* keep quiet about things */
#define MKHOMEDIR_QUIET      040	/* keep quiet about things */

static unsigned int UMask = 0022;
static char SkelDir[BUFSIZ] = "/etc/skel"; /* THIS MODULE IS NOT THREAD SAFE */

/* some syslogging */
static void _log_err(int err, const char *format, ...)
{
    va_list args;

    va_start(args, format);
    openlog("PAM-mkhomedir", LOG_CONS|LOG_PID, LOG_AUTH);
    vsyslog(err, format, args);
    va_end(args);
    closelog();
}

static int _pam_parse(int flags, int argc, const char **argv)
{
   int ctrl = 0;

   /* does the appliction require quiet? */
   if ((flags & PAM_SILENT) == PAM_SILENT)
      ctrl |= MKHOMEDIR_QUIET;

   /* step through arguments */
   for (; argc-- > 0; ++argv)
   {
      if (!strcmp(*argv, "silent")) {
	 ctrl |= MKHOMEDIR_QUIET;
      } else if (!strncmp(*argv,"umask=",6)) {
	 UMask = strtol(*argv+6,0,0);
      } else if (!strncmp(*argv,"skel=",5)) {
	 strncpy(SkelDir,*argv+5,sizeof(SkelDir));
	 SkelDir[sizeof(SkelDir)-1] = '\0';
      } else {
	 _log_err(LOG_ERR, "unknown option; %s", *argv);
      }
   }

   D(("ctrl = %o", ctrl));
   return ctrl;
}

/* This common function is used to send a message to the applications 
   conversion function. Our only use is to ask the application to print 
   an informative message that we are creating a home directory */
static int converse(pam_handle_t * pamh, int ctrl, int nargs
		    ,struct pam_message **message
		    ,struct pam_response **response)
{
   int retval;
   struct pam_conv *conv;

   D(("begin to converse"));

   retval = pam_get_item(pamh, PAM_CONV, (const void **) &conv);
   if (retval == PAM_SUCCESS)
   {

      retval = conv->conv(nargs, (const struct pam_message **) message
			  ,response, conv->appdata_ptr);

      D(("returned from application's conversation function"));

      if (retval != PAM_SUCCESS && (ctrl & MKHOMEDIR_DEBUG))
      {
	 _log_err(LOG_DEBUG, "conversation failure [%s]"
		  ,pam_strerror(pamh, retval));
      }

   }
   else
   {
      _log_err(LOG_ERR, "couldn't obtain coversation function [%s]"
	       ,pam_strerror(pamh, retval));
   }

   D(("ready to return from module conversation"));

   return retval;		/* propagate error status */
}

/* Ask the application to display a short text string for us. */
static int make_remark(pam_handle_t * pamh, int ctrl, const char *remark)
{
   int retval;

   if ((ctrl & MKHOMEDIR_QUIET) != MKHOMEDIR_QUIET)
   {
      struct pam_message msg[1], *mesg[1];
      struct pam_response *resp = NULL;

      mesg[0] = &msg[0];
      msg[0].msg_style = PAM_TEXT_INFO;
      msg[0].msg = remark;

      retval = converse(pamh, ctrl, 1, mesg, &resp);

      msg[0].msg = NULL;
      if (resp)
      {
	 _pam_drop_reply(resp, 1);
      }
   }
   else
   {
      D(("keeping quiet"));
      retval = PAM_SUCCESS;
   }

   D(("returning %s", pam_strerror(pamh, retval)));
   return retval;
}

/* Do the actual work of creating a home dir */
static int create_homedir(pam_handle_t * pamh, int ctrl,
                          const struct passwd *pwd,
                          const char *source, const char *dest)
{
   char remark[BUFSIZ];
   DIR *D;
   struct dirent *Dir;

   /* Mention what is happening, if the notification fails that is OK */
   if (snprintf(remark,sizeof(remark),"Creating directory '%s'.", dest) == -1)
      return PAM_PERM_DENIED;

   make_remark(pamh, ctrl, remark);

   /* Create the new directory */
   if (mkdir(dest,0700) != 0)
   {
      _log_err(LOG_DEBUG, "unable to create directory %s",dest);
      return PAM_PERM_DENIED;
   }   
   if (chmod(dest,0777 & (~UMask)) != 0 ||
       chown(dest,pwd->pw_uid,pwd->pw_gid) != 0)
   {
      _log_err(LOG_DEBUG, "unable to change perms on directory %s",dest);
      return PAM_PERM_DENIED;
   }

   /* See if we need to copy the skel dir over. */
   if ((source == NULL) || (strlen(source) == 0))
   {
      return PAM_SUCCESS;
   }

   /* Scan the directory */
   D = opendir(source);
   if (D == 0)
   {
      _log_err(LOG_DEBUG, "unable to read directory %s",source);
      return PAM_PERM_DENIED;
   }

   for (Dir = readdir(D); Dir != 0; Dir = readdir(D))
   {  
      int SrcFd;
      int DestFd;
      int Res;
      struct stat St;
#ifndef PATH_MAX
      char *newsource=NULL, *newdest=NULL;
      /* track length of buffers */
      int nslen=0, ndlen=0,
	slen=strlen(source), dlen=strlen(dest);
      /* XXX should I check for strlen, sprintf, malloc ? */
#else
      char newsource[PATH_MAX], newdest[PATH_MAX];
#endif

      /* Skip some files.. */
      if (strcmp(Dir->d_name,".") == 0 ||
	  strcmp(Dir->d_name,"..") == 0)
	 continue;

      /* Determine what kind of file it is. */
#ifndef PATH_MAX
      nslen=slen+strlen(Dir->d_name)+2;
      if((newsource=malloc(nslen))==NULL){
	return PAM_BUF_ERR;
      }
      sprintf(newsource,"%s/%s",source,Dir->d_name);
#else
      snprintf(newsource,sizeof(newsource),"%s/%s",source,Dir->d_name);
#endif
      if (lstat(newsource,&St) != 0)
#ifndef PATH_MAX
      {
	 free(newsource);newsource=NULL;
	 continue;
      }
#else
         continue;
#endif


      /* We'll need the new file's name. */
#ifndef PATH_MAX
      ndlen=dlen+strlen(Dir->d_name)+2;
      if((newdest=malloc(ndlen))==NULL){
	free(newsource);
	return PAM_BUF_ERR;
      }
      sprintf(newdest,"%s/%s",dest,Dir->d_name);
#else
      snprintf(newdest,sizeof(newdest),"%s/%s",dest,Dir->d_name);
#endif

      /* If it's a directory, recurse. */
      if (S_ISDIR(St.st_mode))
      {
         create_homedir(pamh, ctrl, pwd, newsource, newdest);
#ifndef PATH_MAX
	 free(newsource);newsource=NULL;
	 free(newdest);newdest=NULL;
#endif
         continue;
      }

      /* If it's a symlink, create a new link. */
      if (S_ISLNK(St.st_mode))
      {
	int pointedlen;
#ifndef PATH_MAX
        char *pointed=NULL;
          {
            int size = 100;
          
            while (1)
              {
                pointed = (char *) malloc (size);
		if(!pointed){
	 	  free(newsource); free(newdest);
		  return PAM_BUF_ERR;
		}
                pointedlen = readlink (newsource, pointed, size);
                if (pointedlen < 0) break;
                if (pointedlen < size) break;
                free (pointed);
                size *= 2;
              }
          }
	 if(pointedlen<0)
	    free(pointed);
	 else
	    pointed[pointedlen]=0;
#else
         char pointed[PATH_MAX];
         memset(pointed, 0, sizeof(pointed));
	 pointedlen=readlink(newsource, pointed, sizeof(pointed) - 1);
#endif
         if(pointedlen>=0) {
            if(symlink(pointed, newdest) == 0)
            {
               if (lchown(newdest,pwd->pw_uid,pwd->pw_gid) != 0)
               {
                   _log_err(LOG_DEBUG, "unable to change perms on link %s",
                            newdest);
#ifndef PATH_MAX
	           free(pointed);
	           free(newsource); free(newdest);
#endif
                   return PAM_PERM_DENIED;
               }
            }
#ifndef PATH_MAX
	   free(pointed);
#endif
         }
#ifndef PATH_MAX
	 free(newsource);newsource=NULL;
	 free(newdest);newdest=NULL;
#endif
         continue;
      }

      /* If it's not a regular file, it's probably not a good idea to create
       * the new device node, FIFO, or whatever it is. */
      if (!S_ISREG(St.st_mode))
      {
#ifndef PATH_MAX
	 free(newsource);newsource=NULL;
	 free(newdest);newdest=NULL;
#endif
         continue;
      }

      /* Open the source file */
      if ((SrcFd = open(newsource,O_RDONLY)) < 0 || fstat(SrcFd,&St) != 0)
      {
         _log_err(LOG_DEBUG, "unable to open src file %s",newsource);
#ifndef PATH_MAX
	 free(newsource); free(newdest);
#endif
	 return PAM_PERM_DENIED;
      }
      stat(newsource,&St);

      /* Open the dest file */
      if ((DestFd = open(newdest,O_WRONLY | O_TRUNC | O_CREAT,0600)) < 0)
      {
	 close(SrcFd);
         _log_err(LOG_DEBUG, "unable to open dest file %s",newdest);
#ifndef PATH_MAX
	 free(newsource); free(newdest);
#endif
	 return PAM_PERM_DENIED;
      }

      /* Set the proper ownership and permissions for the module. We make
       	 the file a+w and then mask it with the set mask. This preseves
       	 execute bits */
      if (fchmod(DestFd,(St.st_mode | 0222) & (~UMask)) != 0 ||
	  fchown(DestFd,pwd->pw_uid,pwd->pw_gid) != 0)
      {
         close(SrcFd);
         close(DestFd);
         _log_err(LOG_DEBUG, "unable to chang perms on copy %s",newdest);
#ifndef PATH_MAX
	 free(newsource); free(newdest);
#endif
	 return PAM_PERM_DENIED;
      }

      /* Copy the file */
      do
      {
         Res = read(SrcFd,remark,sizeof(remark));
	 if (Res < 0 || write(DestFd,remark,Res) != Res)
	 {
	    close(SrcFd);
	    close(DestFd);
	    _log_err(LOG_DEBUG, "unable to perform IO");
#ifndef PATH_MAX
	    free(newsource); free(newdest);
#endif
	    return PAM_PERM_DENIED;
	 }
      }
      while (Res != 0);
      close(SrcFd);
      close(DestFd);
#ifndef PATH_MAX
      free(newsource);newsource=NULL;
      free(newdest);newdest=NULL;
#endif
   }

   return PAM_SUCCESS;
}

/* --- authentication management functions (only) --- */

PAM_EXTERN
int pam_sm_open_session(pam_handle_t * pamh, int flags, int argc
			,const char **argv)
{
   int retval, ctrl;
   const char *user;
   const struct passwd *pwd;
   struct stat St;
      
   /* Parse the flag values */
   ctrl = _pam_parse(flags, argc, argv);

   /* Determine the user name so we can get the home directory */
   retval = pam_get_item(pamh, PAM_USER, (const void **) &user);
   if (retval != PAM_SUCCESS || user == NULL || *user == '\0')
   {
      _log_err(LOG_NOTICE, "user unknown");
      return PAM_USER_UNKNOWN;
   }

   /* Get the password entry */
   pwd = getpwnam(user);
   if (pwd == NULL)
   {
      D(("couldn't identify user %s", user));
      return PAM_CRED_INSUFFICIENT;
   }

   /* Stat the home directory, if something exists then we assume it is
      correct and return a success*/
   if (stat(pwd->pw_dir,&St) == 0)
      return PAM_SUCCESS;

   return create_homedir(pamh,ctrl,pwd,SkelDir,pwd->pw_dir);
}

/* Ignore */
PAM_EXTERN 
int pam_sm_close_session(pam_handle_t * pamh, int flags, int argc
			 ,const char **argv)
{
   return PAM_SUCCESS;
}

#ifdef PAM_STATIC

/* static module data */
struct pam_module _pam_mkhomedir_modstruct =
{
   "pam_mkhomedir",
   NULL,
   NULL,
   NULL,
   pam_sm_open_session,
   pam_sm_close_session,
   NULL,
};

#endif
